src/geometry/manifold/constructions.lean

```diff
@@ -0,0 +1,242 @@
+  convert hf.2.prod_map hg.2 using 1,
+  mfld_set_tac,
+end
+
+lemma smooth_within_at.prod_map
+  (hf : smooth_within_at I I' f s x) (hg : smooth_within_at J J' g t y) :
+  smooth_within_at (I.prod J) (I'.prod J') (prod.map f g) (s.prod t) (x, y) :=
+begin
+  rw times_cont_mdiff_within_at_top at hf hg ⊢,
+  intro n,
+  exact (hf n).prod_map (hg n),
+end
+
+lemma times_cont_mdiff.prod_map {n : with_top ℕ}
+(hf : times_cont_mdiff I I' n f) (hg : times_cont_mdiff J J' n g) :
+  times_cont_mdiff (I.prod J) (I'.prod J') n (prod.map f g) :=
+begin
+  rw ←times_cont_mdiff_on_univ at hf hg ⊢,
+  have h := hf.prod_map hg, rw univ_prod_univ at h,
+  exact h,
+end
+
+lemma times_cont_mdiff_at.prod_map {n : ℕ}
+  (hf : times_cont_mdiff_at I I' n f x) (hg : times_cont_mdiff_at J J' n g y) :
+  times_cont_mdiff_at (I.prod J) (I'.prod J') n (prod.map f g) (x, y) :=
+begin
+  rw times_cont_mdiff_at_iff_times_cont_mdiff_on_nhds at hf hg ⊢,
+  rcases hg with ⟨ug, hug1, hug2⟩, rcases hf with ⟨uf, huf1, huf2⟩,
+  refine ⟨uf.prod ug, prod_mem_nhds_sets huf1 hug1, huf2.prod_map hug2⟩,
+end
+
+lemma smooth_at.prod_map
+  (hf : smooth_at I I' f x) (hg : smooth_at J J' g y) :
+  smooth_at (I.prod J) (I'.prod J') (prod.map f g) (x, y) :=
+by {rw times_cont_mdiff_at_top at hf hg ⊢, intro n, exact (hf n).prod_map (hg n) }
+
+end prod_map
+
+section projections
+
+lemma smooth_fst : smooth (I.prod J) I (@prod.fst M N) :=
+begin
+  rw smooth_iff,
+  refine ⟨continuous_fst, λ x y, _⟩,
+
+  /- I am copying stuff fromt the goal because I do not want to bother spending time to find
+  shorter names, but I'd be happy to have tips on how to find shorter names. -/
+  have h1 := (has_groupoid.compatible (times_cont_diff_groupoid ⊤ (I.prod J))
+    (chart_mem_atlas (H×G) x) (chart_mem_atlas (H×G) (y, x.snd))).1,
+  let s := (prod.map (I.symm) (J.symm) ⁻¹'
+    ((chart_at (model_prod H G) x).to_local_equiv.symm.trans
+    (chart_at (model_prod H G) (y, x.snd)).to_local_equiv).source ∩ set.range (prod.map I J)),
+  have hs : (s ⊆ (λ (x_1 : E × F), (I ((chart_at (model_prod H G) (y, x.snd))
+    (((chart_at (model_prod H G) x).symm) ((I.symm) x_1.fst, (J.symm) x_1.snd))).fst,
+    J ((chart_at (model_prod H G) (y, x.snd)) (((chart_at (model_prod H G) x).symm)
+    ((I.symm) x_1.fst, (J.symm) x_1.snd))).snd)) ⁻¹' (set.univ)) :=
+  by simp only [set.subset_univ, set.preimage_univ],
+  have h2 := times_cont_diff_on.comp (times_cont_diff.times_cont_diff_on times_cont_diff_fst) h1 hs,
+  convert h2 using 1,
+  clear h1 hs h2,
+
+  ext1 z,
+  simp only [set.mem_range, prod_map] with mfld_simps,
+  fsplit,
+  { rintro ⟨⟨⟨⟨a, ha⟩, h1⟩, ⟨b, hb⟩, h2⟩, h3⟩, refine ⟨⟨⟨h1, h3⟩, ⟨h2, _⟩⟩, ⟨a, b⟩, _⟩,
+    { apply local_homeomorph.map_target, /- simp is not working here!!! Why? -/
+      exact h2, },
+    { ext, exacts [ha, hb], } },
+  { rintro ⟨⟨⟨h1, h2⟩, h3, h4⟩, ⟨a, b⟩, rfl⟩, refine ⟨⟨⟨⟨a, rfl⟩, h1⟩, ⟨⟨b, rfl⟩, h3⟩⟩, h2⟩, }
+end
+
+lemma smooth_snd : smooth (I.prod J) J (@prod.snd M N) :=
+begin
```

Comment:
This proof looks really messy. I hope Sébastien will have ideas to streamline it.

Reply:
Yes, most proofs are golfable, but I would rather get things at the right place with the right statements, and then explain to Nicolo some tricks for golfing.